README.md
# rev-static

[![Dependency Status](https://david-dm.org/plantain-00/rev-static.svg)](https://david-dm.org/plantain-00/rev-static)
[![devDependency Status](https://david-dm.org/plantain-00/rev-static/dev-status.svg)](https://david-dm.org/plantain-00/rev-static#info=devDependencies)
[![Build Status: Linux](https://travis-ci.org/plantain-00/rev-static.svg?branch=master)](https://travis-ci.org/plantain-00/rev-static)
[![Build Status: Windows](https://ci.appveyor.com/api/projects/status/github/plantain-00/rev-static?branch=master&svg=true)](https://ci.appveyor.com/project/plantain-00/rev-static/branch/master)
![Github CI](https://github.com/plantain-00/rev-static/workflows/Github%20CI/badge.svg)
[![npm version](https://badge.fury.io/js/rev-static.svg)](https://badge.fury.io/js/rev-static)
[![Downloads](https://img.shields.io/npm/dm/rev-static.svg)](https://www.npmjs.com/package/rev-static)
[![type-coverage](https://img.shields.io/badge/dynamic/json.svg?label=type-coverage&prefix=%E2%89%A5&suffix=%&query=$.typeCoverage.atLeast&uri=https%3A%2F%2Fraw.githubusercontent.com%2Fplantain-00%2Frev-static%2Fmaster%2Fpackage.json)](https://github.com/plantain-00/rev-static)

## features

### add version in file name for css and js files

`index.js`

becomes:

`index-caa02e8ba0c5af68e9ac7728da2bed75.js`

### change file name of css and js files in html files

`<script src="index.js"></script>`

becomes:

`<script src="index-caa02e8ba0c5af68e9ac7728da2bed75.js"></script>`

### calculate sha for css and js files, then add it in `integrity` property

`<script src="index.js" crossOrigin="anonymous"></script>`

becomes:

`<script src="index-caa02e8ba0c5af68e9ac7728da2bed75.js" crossOrigin="anonymous" integrity="sha256-cHLd68M3ekn8P2d8tYdJIV91nSbWrWsu02yI8MEVvYU="></script>`

### support glob

### support config file for options

`rev-static --config rev-static.debug.js`

### inline small js or css file

## install

`yarn global add rev-static`

## usage

```text
rev-static --config rev-static.config.js // config the config file path
rev-static -v
rev-static --version
rev-static --watch
rev-static --config rev-static.config.ts
```

## options

key | description
--- | ---
-w,--watch | watch mode
--config | config file
-h,--help | Print this message.
-v,--version | Print the version

## config file

key | type | use case | description
--- | --- | --- | ---
`inputFiles` | `string[]` | `demo/index.js` | the js, css, html and image file paths, can be glob
`excludeFiles` | `string[]?` | `node_modules/foo/bar.js` | the files will be excluded, can be glob
`revisedFiles` | `string[]?` | `demo/foo-caa02eaf68e9a.js` | the files will be regarded as revised files, can be glob
`inlinedFiles` | `string[]?` | `demo/index.css` | the files will be inlined to html files, can be glob
`outputFiles` | `(file: string) => string` | `demo/index.ejs.html`->`demo/index.html` | the output files mapping function
`ejsOptions` | `EjsOption` | rm whitespace | the options to ejs
`sha` | `256` or `384` or `512` or `undefined` | subresource integrity | the sha type for subresource integrity
`customNewFileName` | `(filePath: string, fileString: string, md5String: string, baseName: string, extensionName: string) => string` | `demo/index.js`->`demo/index-caa02eaf68e9a.js` | the rule of generated js, css, image file
`customOldFileName` | `(filePath: string, baseName: string, extensionName: string) => string` | `demo/index-caa02eaf68e9a.js`->`{ demoIndexJs: demo/index-caa02eaf68e9a.js }` | the rule to get revised file's key
`json` | `string?` | | generate json file for other use
`es6` | `string?` | | generate es6 file for image resources
`less` | `string?` | | generate es6 file for image resources
`scss` | `string?` | | generate es6 file for image resources
`base` | `string?` | | for example, if no `base`, the key will be `demoFooJs` or `demoBarCss`, if the `base` is `demo`, the key will be `fooJs` or `barCss`
`fileSize` | `string?` | | generate a json file to show the resource file size
`context` | `any?` | | will be available when render html template, eg, `context: { foo: 1 }`, then `context.foo === 1`

## core structure

```json
{
  "sri": {
    "fooJs": "sha256-+dZ6udsWxNVoGfScAq7t5IIF5UJb4F6RhjbN6oe1p4w=",
    "foo2Js": "sha256-472ZphjBz2mNWVsknzB8w8rybok1qJl+LKGkqi4XX54=",
    "barCss": "sha256-up7mALOXxS9RzUpTmUIG02GpdK3MsSiMX1Zvco/4J68=",
    "testPng": "sha256-s6pVclbzXpnC5m5oiDf3vdHQskc3AX1vz6h3HRYHgkQ="
  },
  "fooJs": "foo-cb6143ff70a133027139bbf27746a3c4.js",
  "foo2Js": "foo2-42a7b8c.js",
  "barCss": "bar-84f794affa62674553872916601669fe.css",
  "testPng": "test-9d97966727a2d68e865bd7dcf26202df.png"
}
```

## demo in this repository

1. download or clone this repository
1. run `rev-static`(the options is already in `rev-static.config.js`)

## develop

check `scripts` part in `package.json` file, then run `npm run [a script name]`.

## change logs

### v3

+ `noOutputFiles` removed
+ `Usage from nodejs` removed
+ All cli command except `-v`, `--version`, `--config` removed, in favor of config file
+ the `outputFiles` cannot be `string[]` any more

### v2

```js
// before
// default configuration file path: rev-static.config.json

// after
// default configuration file path: rev-static.config.js
```

```js
// before
rev.revisionCssJs(["demo/foo.js", "demo/bar.css"], { shaType: 256 }).then(variables => { });

// after
const variables = rev.revisionCssJs(["demo/foo.js", "demo/bar.css"], { shaType: 256 });
```


CHANGELOG.md
# Change Log

## [3.6.0](https://github.com/plantain-00/rev-st/compare/v3.5.1...v3.6.0) (2020-06-21)
  
* feat: support --help ([32f4d00](https://github.com/plantain-00/rev-st/commit/32f4d00ead6adb182f72df948c7f436add1aac71))
* chore: update dependencies ([3d34176](https://github.com/plantain-00/rev-st/commit/3d34176f37641421fdc506b6694a392990b1c5f2)) ([4b7124f](https://github.com/plantain-00/rev-st/commit/4b7124f7c692fce43c526d276562f063d6579916)) ([d604d98](https://github.com/plantain-00/rev-st/commit/d604d988903d5b301fb405e34cf309bfb731f354)) ([3c8da93](https://github.com/plantain-00/rev-st/commit/3c8da9343473d573ea8d6c094c00366d70ee3189)) ([889e3aa](https://github.com/plantain-00/rev-st/commit/889e3aa061069ecd025bcfa89c367459f3f616f3))

## [3.5.1](https://github.com/plantain-00/rev-st/compare/v3.5.0...v3.5.1) (2020-05-16)
  
* fix: types file ([d66b521](https://github.com/plantain-00/rev-st/commit/d66b5218336ba8654e7322150663f89d2da75618))
* chore: update dependencies ([4d60228](https://github.com/plantain-00/rev-st/commit/4d60228502332716710e55ae0ee03b32304a0a4b))

## [3.5.0](https://github.com/plantain-00/rev-st/compare/v3.4.5...v3.5.0) (2020-05-10)
  
* feat: support typescript config ([dd520f8](https://github.com/plantain-00/rev-st/commit/dd520f8dbe732ec5cbb2887e55d471a085ce1eba))
* chore: update dependencies ([745d3e5](https://github.com/plantain-00/rev-st/commit/745d3e55e8d94354055ec9bdecc92538c9fcd6f2)) ([0964c80](https://github.com/plantain-00/rev-st/commit/0964c806565b8a51b4ff1cd925fa720268e8db70))

## [3.4.5](https://github.com/plantain-00/rev-st/compare/v3.4.4...v3.4.5) (2020-05-01)
  
* chore: loose dependencies version ([841f8e2](https://github.com/plantain-00/rev-st/commit/841f8e2f55fcd94e09d0f84d61042137ce1397fe))
* chore: update dependencies ([3b0d32c](https://github.com/plantain-00/rev-st/commit/3b0d32c73752c15a8347035d3922ffa81c77e8b9)) ([fc94c0f](https://github.com/plantain-00/rev-st/commit/fc94c0f4eeac2fe0a98f2d06a19fe586fb8731f6)) ([0b5a83e](https://github.com/plantain-00/rev-st/commit/0b5a83eee6bb5750991365d1613fb7fb256f8a22)) ([1432562](https://github.com/plantain-00/rev-st/commit/143256255aba0130313b8190e4e6f40e2a10725f)) ([cd9a989](https://github.com/plantain-00/rev-st/commit/cd9a9899d4c38affdff407f2d6b86fee82a8b421)) ([733d37a](https://github.com/plantain-00/rev-st/commit/733d37ada149462607e73b7cbe467223ff861e80))

## [3.4.4](https://github.com/plantain-00/rev-st/compare/v3.4.3...v3.4.4) (2020-03-14)
  

src/index.ts
import * as ejs from 'ejs'
import * as fs from 'fs'
import { createHash } from 'crypto'
import minimist from 'minimist'
import camelcase from 'camelcase'
import * as path from 'path'
import glob from 'glob'
import prettyBytes from 'pretty-bytes'
import minimatch from 'minimatch'
import * as gzipSize from 'gzip-size'
import * as chokidar from 'chokidar'
import * as packageJson from '../package.json'

import { CustomNewFileName, ConfigData, CustomOldFileName } from './core'

function md5(str: string): string {
  return createHash('md5').update(str).digest('hex')
}

function calculateSha(str: string, shaType: 256 | 384 | 512): string {
  return createHash(`sha${shaType}`).update(str).digest('base64')
}

function showToolVersion() {
  console.log(`Version: ${packageJson.version}`)
}

function showHelp() {
  console.log(`Version ${packageJson.version}
Syntax:   rev-static [options]
Examples: rev-static --config rev-static.config.js
          rev-static --config rev-static.config.ts
          rev-static --config rev-static.config.ts --watch
Options:
 -h, --help                                         Print this message.
 -v, --version                                      Print the version
 -w, --watch                                        Watch mode
 --config                                           Config file
`)
}

const htmlExtensions = ['.html', '.htm', '.ejs']

function globAsync(pattern: string, ignore?: string | string[]) {
  return new Promise<string[]>((resolve, reject) => {
    glob(pattern, { ignore }, (error, matches) => {
      if (error) {
        reject(error)
      } else {
        resolve(matches)
      }
    })
  })
}

function renderEjsAsync(filePath: string, data: ejs.Data, opts: ejs.Options) {
  return new Promise<string>((resolve, reject) => {
    ejs.renderFile(filePath, data, opts, (error, file) => {
      if (error) {
        reject(error)
      } else {
        resolve(file)
      }
    })
  })
}

function writeFileAsync(filename: string, data: string) {
  return new Promise<void>((resolve, reject) => {
    fs.writeFile(filename, data, error => {
      if (error) {
        reject(error)
      } else {
        resolve()
      }
    })
  })
}

function getVariableName(filePath: string) {
  return camelcase(path.normalize(filePath).replace(/\\|\//g, '-'))
}

function getNewFileName(fileString: string, filePath: string, customNewFileName?: CustomNewFileName) {
  const md5String = md5(fileString)
  const extensionName = path.extname(filePath)
  const baseName = path.basename(filePath, extensionName)
  if (customNewFileName) {
    return customNewFileName(filePath, fileString, md5String, baseName, extensionName)
  } else {
    return baseName + '-' + md5String + extensionName
  }
}

function getOldFileName(filePath: string, customOldFileName?: CustomOldFileName) {
  const extensionName = path.extname(filePath)
  const baseName = path.basename(filePath, extensionName)
  const dirname = path.dirname(filePath)
  if (customOldFileName) {
    return path.relative('.', path.resolve(dirname, customOldFileName(filePath, baseName, extensionName)))
  } else {
    return path.relative('.', path.resolve(dirname, baseName.split('-')[0] + extensionName))
  }
}

interface Variable {
  name: string;
  value?: string;
  file: string;
  sri?: string;
  fileSize: string;
  inline?: string;
}

function revisionCssOrJs(filePath: string, configData: ConfigData) {
  return new Promise<Variable>((resolve, reject) => {
    fs.readFile(filePath, (error, data) => {
      if (error) {
        reject(error)
      } else {
        const fileString = data.toString()
        let variableName: string
        let newFileName: string | undefined
        const isInlined = configData.inlinedFiles && configData.inlinedFiles.some(inlinedFile => minimatch(filePath, inlinedFile))
        if (configData.revisedFiles
          && configData.revisedFiles.length > 0
          && configData.revisedFiles.some(revisedFile => minimatch(filePath, revisedFile))) {
          const oldFileName = getOldFileName(filePath, configData.customOldFileName)
          variableName = getVariableName(configData.base ? path.relative(configData.base, oldFileName) : oldFileName)
          if (!isInlined) {
            newFileName = path.basename(filePath)
          }
        } else {
          variableName = getVariableName(configData.base ? path.relative(configData.base, filePath) : filePath)
          if (!isInlined) {
            newFileName = getNewFileName(fileString, filePath, configData.customNewFileName)
            fs.createReadStream(filePath).pipe(fs.createWriteStream(path.resolve(path.dirname(filePath), newFileName)))
          }
        }
        const variable: Variable = {
          name: variableName,
          value: newFileName,
          file: filePath,
          fileSize: prettyBytes(fileString.length) + ' ' + prettyBytes(gzipSize.sync(fileString)),
          sri: configData.sha ? `sha${configData.sha}-` + calculateSha(fileString, configData.sha) : undefined,
          inline: undefined
        }
        if (isInlined) {
          if (filePath.endsWith('.js')) {
            variable.inline = `<script>\n${fileString}\n</script>\n`
          } else if (filePath.endsWith('.css')) {
            variable.inline = `<style>\n${fileString}\n</style>\n`
          }
        }
        resolve(variable)
      }
    })
  })
}

function revisionHtml(htmlInputFiles: string[], htmlOutputFiles: string[], variables: Variable[], configData: ConfigData) {
  variables.sort((v1, v2) => v1.name.localeCompare(v2.name))

  const newFileNames: { [name: string]: string } = {}
  const inlineVariables: { [name: string]: string } = {}
  const sriVariables: { [name: string]: string } = {}
  const fileSizes: { [name: string]: string } = {}
  for (const variable of variables) {
    if (variable.value) {
      newFileNames[variable.name] = variable.value
    }
    if (variable.inline) {
      inlineVariables[variable.name] = variable.inline
    }
    if (variable.sri) {
      sriVariables[variable.name] = variable.sri
    }
    fileSizes[variable.name] = variable.fileSize
  }
  const context = Object.assign({ inline: inlineVariables }, { sri: sriVariables }, newFileNames, { context: configData.context })
  const ejsOptions = configData.ejsOptions ? configData.ejsOptions : {}
  Promise.all(htmlInputFiles.map(file => renderEjsAsync(file, context, ejsOptions))).then(fileStrings => {
    for (let i = 0; i < fileStrings.length; i++) {
      const fileString = fileStrings[i]
      writeFileAsync(htmlOutputFiles[i], fileString).then(() => {
        console.log(`Success: to "${htmlOutputFiles[i]}" from "${htmlInputFiles[i]}".`)
      })

      const variableName = getVariableName(htmlOutputFiles[i])
      fileSizes[variableName] = prettyBytes(fileString.length) + ' ' + prettyBytes(gzipSize.sync(fileString))
    }

    if (configData.fileSize) {
      writeFileAsync(configData.fileSize, JSON.stringify(fileSizes, null, '  ')).then(() => {
        console.log(`Success: to "${configData.fileSize}".`)
      })
    }
  })
}

function isImage(key: string) {
  return key !== 'sri' && key !== 'inline' && !key.endsWith('Js') && !key.endsWith('Html') && !key.endsWith('Css')
}

function statAsync(file: string) {
  return new Promise<fs.Stats | undefined>((resolve) => {
    fs.stat(file, (error, stats) => {
      if (error) {
        resolve(undefined)
      } else {
        resolve(stats)
      }
    })
  })
}

async function executeCommandLine() {
  const argv = minimist(process.argv.slice(2), { '--': true }) as unknown as {
    v?: unknown
    version?: unknown
    w?: unknown
    watch?: unknown
    h?: unknown
    help?: unknown
    config?: string
  }

  const showVersion = argv.v || argv.version
  if (showVersion) {
    showToolVersion()
    return
  }

  if (argv.h || argv.help) {
    showHelp()
    return
  }

  let configFilePath: string
  if (argv.config) {
    configFilePath = path.resolve(process.cwd(), argv.config)
  } else {
    configFilePath = path.resolve(process.cwd(), 'rev-static.config.ts')
    const stats = await statAsync(configFilePath)
    if (!stats || !stats.isFile()) {
      configFilePath = path.resolve(process.cwd(), 'rev-static.config.js')
    }
  }
  if (configFilePath.endsWith('.ts')) {
    require('ts-node/register/transpile-only')
  }

  // eslint-disable-next-line @typescript-eslint/no-var-requires
  let configFileData: (ConfigData | ConfigData[]) & { default?: ConfigData | ConfigData[] } = require(configFilePath)
  if (configFileData.default) {
    configFileData = configFileData.default;
  }
  const configDatas = Array.isArray(configFileData) ? configFileData : [configFileData]

  const watchMode = argv.w || argv.watch

  for (const configData of configDatas) {
    if (!configData.inputFiles || configData.inputFiles.length === 0) {
      throw new Error('Error: no input files.')
    }

    const uniqFiles = await globAsync(configData.inputFiles.length === 1 ? configData.inputFiles[0] : `{${configData.inputFiles.join(',')}}`, configData.excludeFiles)
    const htmlInputFiles: string[] = []
    const jsCssInputFiles: string[] = []
    const htmlOutputFiles: string[] = []

    for (const file of uniqFiles) {
      if (isHtmlExtension(file)) {
        htmlInputFiles.push(file)
        htmlOutputFiles.push(configData.outputFiles(file))
      } else {
        jsCssInputFiles.push(file)
      }
    }

    if (watchMode) {
      const variables: Variable[] = []
      let count = 0
      chokidar.watch(configData.inputFiles, { ignored: configData.excludeFiles }).on('all', (type: string, file: string) => {
        console.log(`Detecting ${type}: ${file}`)
        if (type === 'add' || type === 'change') {
          if (isHtmlExtension(file)) {
            const index = htmlInputFiles.findIndex(f => f === file)
            if (index === -1) {
              htmlInputFiles.push(file)
              htmlOutputFiles.push(configData.outputFiles(file))
            } else {
              htmlInputFiles[index] = file
              htmlOutputFiles[index] = configData.outputFiles(file)
            }
            count++
            if (count > uniqFiles.length) {
              revisionHtml(htmlInputFiles, htmlOutputFiles, variables, configData)
            } else if (count === uniqFiles.length) {
              revisionHtml(htmlInputFiles, htmlOutputFiles, variables, configData)
              writeVariables(configData, variables)
            }
          } else {
            revisionCssOrJs(file, configData).then(variable => {
              const index = variables.findIndex(v => v.file === file)
              if (index === -1) {
                variables.push(variable)
              } else {
                const oldVariable = variables[index]
                if (oldVariable.value && oldVariable.value !== variable.value) {
                  const oldFile = path.resolve(path.dirname(oldVariable.file), oldVariable.value)
                  console.log(`Removing ${oldFile}`)
                  fs.unlink(oldFile, error => {
                    if (error) {
                      console.log(error)
                    }
                  })
                }
                variables[index] = variable
              }
              count++

              if (count >= uniqFiles.length) {
                revisionHtml(htmlInputFiles, htmlOutputFiles, variables, configData)
                writeVariables(configData, variables)
              }
            })
          }
        } else if (type === 'unlink') {
          if (isHtmlExtension(file)) {
            const index = htmlInputFiles.findIndex(f => f === file)
            if (index !== -1) {
              htmlInputFiles.splice(index, 1)
              htmlOutputFiles.splice(index, 1)
              revisionHtml(htmlInputFiles, htmlOutputFiles, variables, configData)
            }
          } else {
            const index = variables.findIndex(v => v.file === file)
            if (index !== -1) {
              variables.splice(index, 1)
              revisionCssOrJs(file, configData).then(() => {
                revisionHtml(htmlInputFiles, htmlOutputFiles, variables, configData)
                writeVariables(configData, variables)
              })
            }
          }
        }
      })
    } else if (uniqFiles.length > 0) {
      const variables = await Promise.all(jsCssInputFiles.map(file => revisionCssOrJs(file, configData)))
      revisionHtml(htmlInputFiles, htmlOutputFiles, variables, configData)
      writeVariables(configData, variables)
    }
  }
}

function isHtmlExtension(file: string) {
  return htmlExtensions.indexOf(path.extname(file).toLowerCase()) !== -1
}

function writeVariables(configData: ConfigData, variables: Variable[]) {
  variables.sort((v1, v2) => v1.name.localeCompare(v2.name))

  const newFileNames: { [name: string]: string } = {}
  const fileSizes: { [name: string]: string } = {}
  for (const variable of variables) {
    if (variable.value) {
      newFileNames[variable.name] = variable.value
    }
    fileSizes[variable.name] = variable.fileSize
  }

  if (configData.json) {
    writeFileAsync(configData.json, JSON.stringify(newFileNames, null, '  ')).then(() => {
      console.log(`Success: to "${configData.json}".`)
    })
  } else {
    console.log(`File Renamed To: ${JSON.stringify(newFileNames, null, '  ')}`)
  }

  if (configData.es6) {
    const result: string[] = []
    for (const variable of variables) {
      if (isImage(variable.name) && variable.value !== undefined) {
        result.push(`export const ${variable.name} = "${variable.value}";\n`)
      }
    }

    writeFileAsync(configData.es6, result.join('')).then(() => {
      console.log(`Success: to "${configData.es6}".`)
    })
  }

  if (configData.less) {
    const result: string[] = []
    for (const variable of variables) {
      if (isImage(variable.name) && variable.value !== undefined) {
        result.push(`@${variable.name}: '${variable.value}';\n`)
      }
    }

    writeFileAsync(configData.less, result.join('')).then(() => {
      console.log(`Success: to "${configData.less}".`)
    })
  }

  if (configData.scss) {
    const result: string[] = []
    for (const variable of variables) {
      if (isImage(variable.name) && variable.value !== undefined) {
        result.push(`$${variable.name}: '${variable.value}';\n`)
      }
    }

    writeFileAsync(configData.scss, result.join('')).then(() => {
      console.log(`Success: to "${configData.scss}".`)
    })
  }
}

executeCommandLine().then(() => {
  console.log('rev-static success.')
}, error => {
  if (error instanceof Error) {
    console.log(error.message)
  } else {
    console.log(error)
  }
  process.exit(1)
})
